src/librustc/traits/error_reporting.rs

```diff
@@ -837,6 +832,47 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
         err
     }
+
+    fn report_closure_arg_mismatch(&self,
+                           span: Span,
+                           found_span: Option<Span>,
+                           expected_ref: ty::PolyTraitRef<'tcx>,
+                           found: ty::PolyTraitRef<'tcx>)
+        -> DiagnosticBuilder<'tcx>
+    {
+        fn build_args_type_string(trait_ref: &ty::TraitRef) -> String {
+            match trait_ref.substs.type_at(1).sty {
+                ty::TyTuple(args, _) => {
+                    let len = args.len();
+                    if len == 0 {
+                        return String::from("no arguments");
```

Comment:
Is this code unreachable? I mean, will `report_closure_arg_mismatch` be called even when one of the `Fn` traits has zero arguments?

Reply:
Why aren't you calling `report_type_argument_mismatch`?

Reply:
ok I see

Reply:
I think you won't see this error message - no arguments, no place for a type error.